index/scorch/persister.go

```diff
@@ -267,17 +267,36 @@ func (s *Scorch) pausePersisterForMergerCatchUp(lastPersistedEpoch uint64,
 	// memory merge cum persist loop.
 	if numFilesOnDisk < uint64(po.PersisterNapUnderNumFiles) &&
 		po.PersisterNapTimeMSec > 0 && s.NumEventsBlocking() == 0 {
+		s.rootLock.RLock()
+		lastRootEpoch := s.root.epoch
+		s.rootLock.RUnlock()
+		expectedNapTime := time.Now().Add(time.Millisecond * time.Duration(po.PersisterNapTimeMSec))
+
 		select {
 		case <-s.closeCh:
 		case <-time.After(time.Millisecond * time.Duration(po.PersisterNapTimeMSec)):
 			atomic.AddUint64(&s.stats.TotPersisterNapPauseCompleted, 1)
 
 		case ew := <-s.persisterNotifier:
+			s.rootLock.RLock()
+			currRootEpoch := s.root.epoch
```

Comment:
@Thejas-bhat Shouldn't we be checking the epoch from the persisterNotifier `epochWatcher` instead or somewhere?

Reply:
I understand why we're comparing this epoch here, but is there any possibility of a race here -

1. the merger updates the root epoch, sets up an epoch watcher
2. the persister then records the lastRootEpoch in line 271 above
3. the comparison in line 289 below passes because no other change since, and the persister sleeps longer than it should have.

Reply:
from what i understand that shouldn't happen, this is because
1. when the merger has some changes to introduce, the introducer will get the message of the newly merged segments to be introduced. the introducer now updates the root epoch after acquiring a lock. after this, the merger notifies the persister. so the updating of the rootEpoch followed by notifying the persister is happening in a sequential manner.
2. the lastRootEpoch value at 271 is still the pre-merged snapshot and now at 289 when we get the currRootEpoch at that instance using the read lock, the currRootEpoch value would be the new root epoch - so the check would fail and the persister continues to do the work